[CelestineTan03] iP

CelestineTan03 opened the pull request:
## DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,
1. download the duke.jar file from [here](https://github.com/CelestineTan03/ip/releases/tag/A-Jar).
2. run `java -jar duke.jar`.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features: 

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

CelestineTan03 committed:
Rename chatbot and make chatbot greet and exit

CelestineTan03 committed:
Chatbot now echos user's command

CelestineTan03 committed:
Add list

CelestineTan03 committed:
Add the ability to mark and unmark tasks

CelestineTan03 committed:
Add support for tracking ToDos, Deadlines and Events

CelestineTan03 committed:
Test using the I/O redirection technique

CelestineTan03 committed:
Add input validation

CelestineTan03 committed:
Add support for deleting tasks from the list

CelestineTan03 committed:
Update error handling and add javadoc

CelestineTan03 committed:
feat: add the ability to save data

CelestineTan03 committed:
Add date and time feature

CelestineTan03 committed:
refactor: extract Ui, Storage, Parser, TaskList classes

CelestineTan03 committed:
Add package

CelestineTan03 committed:
Merge branch 'add-gradle-support'

CelestineTan03 committed:
Add JUnit tests

CelestineTan03 committed:
Create JAR file

CelestineTan03 committed:
Add JavaDoc comments

CelestineTan03 committed:
Update code to comply with coding standard

CelestineTan03 committed:
feat: find task by searching for keyword

CelestineTan03 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Task.java

CelestineTan03 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/Todo.java

CelestineTan03 committed:
fix: text-ui-test fail due to previous test data stored and input to wrong program

CelestineTan03 committed:
Add checkstyle

CelestineTan03 committed:
Add a GUI to the chatbot

bnahor commented on src/main/java/duke/Duke.java:
```diff
@@ -25,110 +26,89 @@ public Duke(String filePath) {
-                    this.ui.printAllCommands();
-                    break;
-                }
+        switch (firstWord) {
```
I like the way you've structured this switch case statement; all of the exceptions are caught and handled, and most of the parsing logic is passed off to the relevant classes!

bnahor commented on src/main/java/duke/Ui.java:
```diff
@@ -4,29 +4,27 @@
-        System.out.println("> delete <task number>");
-        System.out.println("> find <keyword>");
-        System.out.println("> bye");
+    public String printAllCommands() {
```
Consider using String.join() instead?

bnahor reviewed (commented):
LGTM, some finishing touches

kokrui commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class Ui {
+
+    // Print the farewell message
```
comments above methods should be JavaDoc comments, i.e.

```suggestion
    /**
    * Prints the farewell message
    */
```

kokrui commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,63 @@
+     * @return the status icon of the task
+     */
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
Instead of using magic strings like "X" and " ", one could consider declaring them as class level constants with meaningful names or use enums

kokrui commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,209 @@
+     * Takes in the task details and add the task to the list.
+     *
+     * @param task the task description
+     * @param typeOfTask the type of the task to be added
```
One could consider renaming `typeOfTask` to just `taskType`, which is simpler without losing its descriptive quality

kokrui reviewed (commented):
Really clean and well-structured code overall -- suggested some minor code quality nits

erohsikivar commented on src/main/java/duke/ExceptionTypes.java:
```diff
@@ -0,0 +1,37 @@
+    DELETEEMPTYLIST("Add task to start deleting!"),
+    INVALIDCOMMAND("OOPS!!! I'm sorry, but I don't know what that means :-("),
+    EMPTYKEYWORD("OOPS!!! Please enter something to search.");
+
```
Good to see that you have used enums which I did not use.

erohsikivar commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,209 @@
+/**
+ * Represents a list of tasks.
+ */
+public class TaskList {
```
Good to see that you have different methods for each command but you can also use only one switch - case statements for all methods instead.

erohsikivar reviewed (commented):
Overall LGTM, you can also consider using more than one package even though it is not necessary.

CelestineTan03 committed:
Generate new JAR file

CelestineTan03 committed:
Add assertions

CelestineTan03 committed:
Improve code quality

Javadoc comments are not following the Java coding standard.

Let's,
- add punctuation behind every parameter description
- start the first sentence in method header comments
in the form "Returns...", "Adds..."

CelestineTan03 committed:
Use Streams

Tasks are filtered using for loop.

Let's filter tasks using stream.

CelestineTan03 committed:
Merge pull request #1 from CelestineTan03/Branch-A-Assertions

Add assertions

CelestineTan03 committed:
Merge pull request #2 from CelestineTan03/Branch-A-CodeQuality

Improve code quality

Javadoc comments are not following the Java coding standard.

Let's,
- add punctuation behind every parameter description
- start the first sentence in method header comments in the form "Returns...", "Adds..."

CelestineTan03 committed:
Merge pull request #3 from CelestineTan03/Branch-A-Streams

Use Streams

Tasks are filtered using for loop.

Let's filter tasks using stream.

CelestineTan03 committed:
Set up CI

CelestineTan03 committed:
feat: update task

Users can only add or delete task.

It will be easier if users can edit the existing task if there are any changes.

Let's add an update task feature.

CelestineTan03 committed:
fix: incorrect assertion

CelestineTan03 committed:
Update CI

CelestineTan03 committed:
Improve GUI

CelestineTan03 committed:
Add product name and representative screenshot of product

CelestineTan03 committed:
Add user guide

CelestineTan03 committed:
Release the product

CelestineTan03 committed:
Refactor long method

CelestineTan03 committed:
Add javadoc

CelestineTan03 committed:
Give a personality

CelestineTan03 committed:
Update ParserTest
